File: rust/day01/src/main.rs
use std::fs;
use std::str::FromStr;

type DepthMeasurment = usize;

fn parse_input_file(path: &str) -> Vec<DepthMeasurment> {
    let contents = fs::read_to_string(path).unwrap();
    let measurement = contents.lines().map(|line| usize::from_str(line).unwrap());
    measurement.collect()
}

fn count_increases(measurements: &[DepthMeasurment]) -> usize {
    measurements.windows(2).filter(|pair| pair[1] > pair[0]).count()
}

fn count_triple_increases(measurements: &[DepthMeasurment]) -> usize {
    measurements.windows(4).filter(|values| values[3] > values[0]).count()
}

fn main() {
    let measurements = parse_input_file("input.txt");
    let depth_increases = count_increases(&measurements);
    println!("Answer 1: {}", depth_increases);
    let triple_depth_increases = count_triple_increases(&measurements);
    println!("Answer 2: {}", triple_depth_increases);
}

File: haskell/day01/app/Main.hs
module Main where

import System.IO as IO

type DepthMeasurement = Word

parseInputFile :: FilePath -> IO [Word]
parseInputFile path = do
    contents <- IO.readFile path
    return $ map read $ lines contents

countIncreases :: [DepthMeasurement] -> Word
countIncreases (x1:x2:xs) =
    if x2 > x1
    then 1 + recurse
    else recurse
  where
    recurse = countIncreases (x2:xs)
countIncreases _ = 0

countIncreasesTriple :: [DepthMeasurement] -> Word
countIncreasesTriple (x1:x2:x3:x4:xs) =
    if x4 + x3 + x2 > x3 + x2 + x1
    then 1 + recurse
    else recurse
  where
    recurse = countIncreasesTriple (x2:x3:x4:xs)
countIncreasesTriple _ = 0

main :: IO ()
main = do
    measurements <- parseInputFile "input.txt"
    let depthIncreasesCount = countIncreases measurements
    putStrLn $ "Answer 1: " <> show depthIncreasesCount
    let depthIncreasesTripleCount = countIncreasesTriple measurements
    putStrLn $ "Answer 2: " <> show depthIncreasesTripleCount
